File: 1340-jump-game-v/1340-jump-game-v.cpp
class Solution {
public:

    int n;
    vector<int>dp;

    int solution(int i,vector<int>& arr, int d){
        if(dp[i]!=-1) return dp[i];
        int ans=1;
        for(int j=i+1;j<=i+d && j<n;j++){
            if(arr[j]>=arr[i]) break;
            ans=max(ans,1+solution(j,arr,d));
        }

        for(int j=i-1;j>=i-d && j>=0;j--){
            if(arr[j]>=arr[i]) break;
            ans=max(ans,1+solution(j,arr,d));
        }
        return dp[i]=ans;
    }

    int maxJumps(vector<int>& arr, int d) {
        n=arr.size();
        dp.assign(n,-1);
        int maxi=INT_MIN;
        for(int i=0;i<n;i++){
            maxi=max(maxi,solution(i,arr,d));
        }
        return maxi;
    }
};

File: 1340-jump-game-v/1340-jump-game-v.java
class Solution {

    int n;
    int dp[];

    int solution(int i,int arr[],int d){
        if(dp[i]!=-1) return dp[i];
        int ans=1;

        for(int j=i+1;j<n && j<=i+d;j++){
            if(arr[j]>=arr[i]) break;
            ans=Math.max(ans,1+solution(j,arr,d));
        }

        for(int j=i-1;j>=0 && j>=i-d;j--){
            if(arr[j]>=arr[i]) break;
            ans=Math.max(ans,1+solution(j,arr,d));
        }
        return dp[i]=ans;
    }

    public int maxJumps(int[] arr, int d) {
        n=arr.length;
        dp=new int[n];
        Arrays.fill(dp,-1);
        int maxi=0;
        for(int i=0;i<n;i++){
            maxi=Math.max(maxi,solution(i,arr,d));
        }
        return maxi;
    }
}